cpp/marc_tools/single_marc_convert_keywords.cc

```diff
@@ -0,0 +1,145 @@
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <iostream>
+#include <set>
+#include <iterator>
+#include <utility>
+#include <stdexcept>
+#include <climits>
+#include <cstdio>
+#include <cstdlib>
+#include <fstream>
+#include <iomanip>
+#include "FileUtil.h"
+#include "MARC.h"
+#include "StringUtil.h"
+#include "util.h"
+
+
+namespace {
+
+
+[[noreturn]] void Usage() {
+    ::Usage("gnd_input keyword_input keyword_matches_output keyword_without_matches_output"
+               "       Searches for keyword matches in the \"gnd_input\" file."
+               "       Returns a \"keyword_matches_output\" file with matching keywords and their PPN,"
+               "       as well as \"keywords_without_matches\" file containing keywords where no matches were found.\n");
+}
+
+
+void ReadInGndKeywords(MARC::Reader * const marc_reader, std::unordered_map<std::string, std::string> * const gnd_keywords)
+{
+    unsigned record_count(0);
+
+    while (MARC::Record record = marc_reader->read()) {
+        ++record_count;
+
+        if (not record.getFirstSubfieldValue("150", 'a').empty())
+            gnd_keywords->emplace(std::make_pair(record.getFirstSubfieldValue("150", 'a'), record.getControlNumber()));
+        if (not record.getFirstSubfieldValue("150", 'g').empty())
+            gnd_keywords->emplace(std::make_pair(record.getFirstSubfieldValue("150", 'g'), record.getControlNumber()));
+        if (not record.getFirstSubfieldValue("150", 'x').empty())
+            gnd_keywords->emplace(std::make_pair(record.getFirstSubfieldValue("150", 'x'), record.getControlNumber()));
+    }
+
+    LOG_INFO("Processed " + std::to_string(record_count) + " MARC record(s).");
+    
+}
+
+
+void FindEquivalentKeywords(std::unordered_map<std::string, std::string> const *keywords_to_gnd, std::unordered_set<std::string> const *keywords_to_compare, const std::string matches_output_file, const std::string no_matches_output_file) {
+    std::unordered_map<std::string, std::string> keyword_matches;
+    std::unordered_set<std::string> keywords_without_match;
+    for (const auto &keyword : *keywords_to_compare) {
+        const auto lookup(keywords_to_gnd->find(keyword));
+        if (lookup != keywords_to_gnd->end()) {
+            keyword_matches.insert(*lookup);
+            continue;
+        }
+        keywords_without_match.insert(keyword);
+    }
+    LOG_INFO("Found " + std::to_string(keyword_matches.size()) + " keyword matches.\n");
+    double percentage = (static_cast<double>(keyword_matches.size())/static_cast<double>(keywords_to_compare->size())) * 100;
```

Comment:
percentage should be const, also we use parentheses for variable initialisation and not equal signs.